### PR TITLE
Spring Boot Security implementation

### PR DESCRIPTION
**Pull request by ADIGrimm**

### CODE REVIEW

File: src/main/java/online/bookstore/controller/BookController.java [hunk 1]
```diff
@@ -26,32 +27,38 @@ public class BookController {
     private final BookService bookService;
 
     @GetMapping
+    @PreAuthorize("hasAnyRole('ROLE_USER', 'ROLE_ADMIN')")
```
Comment (liliia-ponomarenko) [inline src/main/java/online/bookstore/controller/BookController.java]: In your solution, the user-role has many-to-many relations. You give a USER role by default to each registered user, meaning that ADMIN also has a USER role. So you can use only the USER role here  
```suggestion
    @PreAuthorize("hasAnyRole('ROLE_USER')")
```

File: src/main/java/online/bookstore/controller/BookController.java [hunk 2]
```diff
@@ -26,32 +27,38 @@ public class BookController {
...
     public List<BookDto> getAll(Pageable pageable) {
         return bookService.findAll(pageable);
     }
 
     @GetMapping("/{id}")
+    @PreAuthorize("hasAnyRole('ROLE_USER', 'ROLE_ADMIN')")
```
Comment (liliia-ponomarenko) [inline src/main/java/online/bookstore/controller/BookController.java]: the same (check other endpoint’s)

File: src/main/java/online/bookstore/model/Role.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    public String getAuthority() {
+        return name.name();
+    }
+
+    public enum RoleName {
+        ROLE_USER,
+        ROLE_ADMIN
+    }
+}
+
```
Comment (liliia-ponomarenko) [inline src/main/java/online/bookstore/model/Role.java]:
```suggestion
```
only one empty line

File: src/main/resources/db/changelog/changes/07-create-association-role-for-admin.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+      changes:
+        - insert:
+            tableName: users_roles
+            columns:
+              - column:
+                  name: user_id
+                  value: 1
+              - column:
+                  name: role_id
+                  value: 2
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/07-create-association-role-for-admin.yaml]: add one empty line

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! A few minor comments ;)

File: src/main/java/online/bookstore/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -1,23 +1,34 @@
...
+    private final PasswordEncoder passwordEncoder;
+    private final RoleRepository roleRepository;
 
     @Override
+    @Transactional
     public UserResponseDto registration(UserRegistrationRequestDto requestDto) {
         User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        user.setRoles(Set.of(roleRepository.findByName(Role.RoleName.ROLE_USER)
+                .orElseThrow(() -> new RuntimeException("Can't find role"))));
```
Comment (Elena-Bruyako) [inline src/main/java/online/bookstore/service/impl/UserServiceImpl.java]: wrong exception

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
one comment

File: src/main/java/online/bookstore/service/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -1,23 +1,35 @@
...
+    private final PasswordEncoder passwordEncoder;
+    private final RoleRepository roleRepository;
 
     @Override
+    @Transactional
     public UserResponseDto registration(UserRegistrationRequestDto requestDto) {
         User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        user.setRoles(Set.of(roleRepository.findByName(Role.RoleName.ROLE_USER)
+                .orElseThrow(() -> new RegistrationException("Can't find role"))));
```
Comment (Elena-Bruyako) [inline src/main/java/online/bookstore/service/impl/UserServiceImpl.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException("Can't find role"))));
```

File: src/main/java/online/bookstore/service/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -1,23 +1,35 @@
...
 @RequiredArgsConstructor
 public class UserServiceImpl implements UserService {
     private final UserRepository userRepository;
     private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
+    private final RoleRepository roleRepository;
 
     @Override
+    @Transactional
     public UserResponseDto registration(UserRegistrationRequestDto requestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/online/bookstore/service/impl/UserServiceImpl.java:29]:
```suggestion
    public UserResponseDto registration(UserRegistrationRequestDto requestDto) {
        if (userRepo.existsByEmail(request.email())) {
            throw new RegistrationException(
                    String.format("User with this email: %s already exists", request.email())
            );
        }
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

**Review by ahoienko — APPROVED** [review]
Good job!

**Pull request merged**